hourly-tips/src/main/java/org/apache/flink/training/exercises/hourlytips/HourlyTipsExercise.java

```diff
@@ -72,22 +80,46 @@ public JobExecutionResult execute() throws Exception {
         // set up streaming execution environment
         StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
 
-        // start the data generator
-        DataStream<TaxiFare> fares = env.addSource(source);
+        WatermarkStrategy<TaxiFare> strategy
+                = WatermarkStrategy.<TaxiFare>forMonotonousTimestamps()
+                .withTimestampAssigner((fare, assigner) -> fare.getEventTimeMillis());
 
-        // replace this with your solution
-        if (true) {
-            throw new MissingSolutionException();
-        }
+        DataStream<TaxiFare> fares = env.addSource(source).assignTimestampsAndWatermarks(strategy);
 
-        // the results should be sent to the sink that was passed in
-        // (otherwise the tests won't work)
-        // you can end the pipeline with something like this:
+        // the pipeline:
 
-        // DataStream<Tuple3<Long, Long, Float>> hourlyMax = ...
-        // hourlyMax.addSink(sink);
+        fares   .keyBy((TaxiFare fare) -> fare.driverId)
+                .window(TumblingEventTimeWindows.of(windowDuration))
+                .process(new SumWindowTips())
 
-        // execute the pipeline and return the result
-        return env.execute("Hourly Tips");
+                .windowAll(TumblingEventTimeWindows.of(windowDuration))  // non-key grouped stream
+                .maxBy(2) // max by sumOfTips (f2 in Tuple3)
+
+                .addSink(this.sink)
+                ;
+
+        return env.execute("Hourly Tips"); // execute the pipeline and return the result
+    }
+
+    public static class SumWindowTips
+            extends ProcessWindowFunction<TaxiFare, Tuple3<Long, Long, Float>, Long, TimeWindow> {
+        @Override
+        public void process
```

Comment:
Sorry, but I can read this Method because of bad code formatting.
Please apply proper code formatting (default one from IntelliJ IDEA).

Reply:
ok. done